### DISCUSSION TITLE
A solution for dark combo colours

### DISCUSSION THREAD

**Discussion by TheFantasticLoki** [ppy/osu#19682] [Ideas]
Something that has always irritated me as a long time player and multiplayer main especially is black or really dark combo colors in random maps even some ranked maps, therefore I would LOVE to see the addition of a setting in both lazer and stable that allows you to set a minimum total rgb value or similar. What I mean is the total of all r g and b channels values added together must be above a set value otherwise they are multiplied by a small amount till the minimum total is reached in order to keep the color ratio the same but overall brighten extremely dark combo colors in maps. That is just 1 approach but the one I like the most as a default but another alternative would be the same concept but instead of multiplying rgb values just setting a user defined combo color range. 

My Main overall reason for requesting this addition INSTEAD of just using ignore beatmap skin in stable and ignore beatmap combo colors in lazer is because while these do exist and work as intended they must either be for all maps or toggled per map AFTER discovering the issue on the sightread play therefore ruining many sightread plays of not only myself by my friends and many other players I have met over my years of multiplayer. Many maps are hand crafted with combo colors in mind and disabling that by default for every map because random maps can become unplayable in multi just isn't good enough in my opinion.

Figure you would want some examples of the issue so here are some screenshots of various maps with dark combo colors causing visibility issues:
In this screenshot you can see an example of black combo colors and a very dark blue in use on a very fun stream map I play frequently, This is one of those cases where a black combo color is especially bad since streaming relies on approach circles quite heavily.
![screenshot730](https://user-images.githubusercontent.com/26889568/183871335-05a5f0f6-93b6-4302-a27c-a56e8a747a51.png)

Another inherent flaw of these dark combo colors is that in certain skins especially any that use only hitcircle in the skin with no hitcircleoverlay, circles can appear almost invisible on screen. In my skin my sliders use this technique to further distinquish sliders from normal hitcircles and the slider heads with black combo color are very hard to see. I also play with 93% bg dim by default but most play with 100 further adding to this problem:
![screenshot732](https://user-images.githubusercontent.com/26889568/183872436-72007333-4167-44e4-8100-1904dbf7df3e.png)

Another example I forgot to use is this color matched combo color scheme on a loved map that causes reading and visibility issues:
![screenshot725](https://user-images.githubusercontent.com/26889568/183872783-b63f7cd1-a69e-4189-831a-9fd98d23553e.png)
![screenshot726](https://user-images.githubusercontent.com/26889568/183872798-c1965a58-07a3-45a9-9521-f9c6444e39a0.png)

**Comment by nagi-desuuu** [comment]
Stable is feature-locked so this won't happen there.
However, you can easily remedy this by disabling "Beatmap colors"

**Comment by smoogipoo** [comment]
I suppose we could convert combo colours to HSP and limit the brightness. Would be curious to see how that works in practice.

**Reply by TheFantasticLoki** [reply]
Would this work by just limiting the max brightness or both max and min? I'm not familiar with how HSP works but based off what you have typed I have a feeling it will only limit the max brightness which hasn't been an issue.

**Reply by smoogipoo** [reply]
Both.

**Reply by smoogipoo** [reply]
Fwiw I've tried an implementation of this proposal here: https://github.com/smoogipoo/osu/tree/combo-colour-brightness-limit

I've made it pretty aggressive for demonstration purposes but here's how it looks:
| master | this branch |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/1329837/194021513-3cb72238-6e30-47e1-92f7-18b2ad30357e.png) | ![image](https://user-images.githubusercontent.com/1329837/194021238-03fdcccd-5fa1-4f77-ae15-87489a9c7242.png) |

Of importance, the conversion process loses data. I'm not sure if this is a bug with my code or if it's inherent to the algorithm. It may work well applied unconditionally to all combo colours instead, as a "normalisation" pass - from my initial testing this looks good, but as I said it loses some colour data.

**Reply by smoogipoo** [reply]
And another comparison showing the colour data being lost (if applied unconditionally):

Before:

https://user-images.githubusercontent.com/1329837/194025322-25cb607c-ce93-44a3-aaf9-4d0a69d4e734.mp4

After:

https://user-images.githubusercontent.com/1329837/194025344-7840f15b-8734-4f20-b772-c8d62e77a16b.mp4

**Reply by smoogipoo** [reply]
Looks like the loss in colour was my error (keep in mind this comparison is at 50% brightness):

https://user-images.githubusercontent.com/1329837/194026795-68a2dcad-ef50-4a5e-b580-d6d4048f6f6c.mp4

Shall I try to PR this somehow? @peppy

**Reply by peppy** [reply]
I think PRing is a good place to start. I'd likely want to tweak things and test further.

**Reply by goodtrailer** [reply]
would this be a per-map option or depend on background darkness (post-dim)? some maps just have really bright backgrounds so use dark colors for contrast

**Reply by peppy** [reply]
it would be applied without user interaction, and not using background colour. it shouldn't need either. if it does then it can't exist.
